### PR TITLE
adding string casting for non-bytes values

### PR DESCRIPTION
**Pull request by nateprewitt**
This addresses #3753, I'll open a PR for 3.0.0 removing this when we merge master.

### CODE REVIEW

**Comment by kennethreitz** [comment]
i support this

**Comment by kennethreitz** [comment]
it's stupid, but no reason not to

**Comment by Lukasa** [comment]
LGTM. Thanks for this @nateprewitt! :sparkles:

**Comment by Lukasa** [comment]
And I've just caught the 3.0.0 branch up, so you can go ahead and open the second PR now.

File: requests/auth.py [hunk 1]
```diff
@@ -27,7 +27,21 @@
...
-    
+
+    # "I want us to put a big-ol' comment on top of it that
+    # says that this behaviour is dumb but we need to preserve
+    # it because people are relying on it."
+    #    - Lukasa
+    #
+    # These are here solely to maintain backwards compatibility
+    # for things like ints. This will be removed in 3.0.0.
+    if not isinstance(username, basestring):
```
Comment (sigmavirus24) [inline requests/auth.py:38]: We should also have included a deprecation warning in here too to warn people this was happening in 3.0

File: tests/test_requests.py [hunk 2]
```diff
@@ -484,6 +484,7 @@ def test_BASICAUTH_TUPLE_HTTP_200_OK_GET(self, httpbin):
         'username, password', (
             ('user', 'pass'),
             (u'имя'.encode('utf-8'), u'пароль'.encode('utf-8')),
+            (42, 42)
```
Comment (sigmavirus24) [inline tests/test_requests.py:487]: This should also have had a test with `(None, None)` just to cover all of our bases.

File: requests/auth.py (hunk 1 above)
Comment (Lukasa) [reply requests/auth.py:38]: Honestly, a whole lot of stuff got removed that doesn't have deprecation warnings. I kinda think that ship has sailed.

Comment (sigmavirus24) [reply requests/auth.py:38]: The difference being that we're adding this *back* with the intent to remove it. If we're doing that, we have enough time to add a deprecation warning

Comment (Lukasa) [reply requests/auth.py:38]: I'm certainly happy to accept a patch that adds the deprecation warning. =) I have no objection to adding it.

**Comment by nateprewitt** [comment]
@sigmavirus24 you're suggesting using `warnings.warn` to actively notify the user about the deprecation, correct? I can add that and the test here in a moment if you're not already working on it.
